runtime/arm64.S

```diff
@@ -610,8 +616,28 @@ FUNCTION(caml_c_call_stack_args)
     /* Make the exception handler alloc ptr available to the C code */
         str     ALLOC_PTR, Caml_state(young_ptr)
         str     TRAP_PTR, Caml_state(exn_handler)
-    /* Store sp to restore after call */
-        mov     x19, sp
+    /* Copy the arguments and call */
+        bl      G(caml_c_call_copy_stack_args)
+    /* Reload new allocation pointer & exn handler */
+        ldr     ALLOC_PTR, Caml_state(young_ptr)
+        ldr     TRAP_PTR, Caml_state(exn_handler)
+    /* Switch from C to OCaml */
+        SWITCH_C_TO_OCAML
+    /* Return */
+        LEAVE_FUNCTION
+        RET_FROM_C_CALL
+        CFI_ENDPROC
+END_FUNCTION(caml_c_call_stack_args)
+
+/* To correctly maintain frame pointers during stack reallocation,
+   the runtime assumes that the caml_c_call stub does not push
+   anything to the stack before the first frame pointer on the C stack.
+   To guarantee this when stack arguments are used, the actual pushing
+   of arguments is done by this separate function */
+FUNCTION(caml_c_call_copy_stack_args)
+        CFI_STARTPROC
```

Comment:
Not labelling this with `CFI_SIGNAL_FRAME` will mean gdb shows this as a stack frame rather than `<signal handler called>`.  On ARM64 Linux running the `c_call.ml` test:

```shell
(gdb) bt
#0  0x0000aaaaaaae8078 in caml_c_call_copy_stack_args ()
#1  <signal handler called>
#2  0x0000aaaaaaabc618 in camlC_call.f_274 () at c_call.ml:16
#3  0x0000aaaaaaabc6e8 in camlC_call.entry () at c_call.ml:24
#4  0x0000aaaaaaabafdc in caml_program ()
#5  <signal handler called>
#6  0x0000aaaaaaae7b88 in caml_startup_common (pooling=-1430997904, argv=0xaaaaaab4b460) at runtime/startup_nat.c:127
```

macOS LLDB doesn't care and will show the full backtrace, using the same test case.
```shell
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
  * frame #0: 0x0000000100008a00 c_call.opt`fp_backtrace_many_args(argv0=4302438272, a=3, b=5, c=7, d=9, e=11, f=13, g=15, h=17, i=19, j=21, k=23) at c_call_.c:24:3 [opt]
    frame #1: 0x00000001000391d0 c_call.opt`caml_c_call_copy_stack_args + 36
    frame #2: 0x0000000100039180 c_call.opt`caml_c_call_stack_args + 52
    frame #3: 0x0000000100003a98 c_call.opt`camlC_call.f_274 + 128
    frame #4: 0x0000000100003b68 c_call.opt`camlC_call.entry + 72
    frame #5: 0x000000010000344c c_call.opt`caml_program + 116
    frame #6: 0x0000000100039260 c_call.opt`caml_start_program + 132
```

I noticed the stack trace gets broken when you enter `fp_backtrace_many_args` on ARM64 GDB:
```shell
(gdb) bt
#0  fp_backtrace_many_args (argv0=281474571960208, a=<optimized out>, b=5, c=7, d=9, e=11, f=13, g=15, h=17, i=19, 
    j=21, k=23) at c_call_.c:25
#1  0x0000aaaaaaae8090 in caml_c_call_copy_stack_args ()
#2  0x0000aaaaaab4b450 in ?? ()
Backtrace stopped: previous frame inner to this frame (corrupt stack?)
(gdb) info breakpoints 
Num     Type           Disp Enb Address            What
1       breakpoint     keep y   0x0000aaaaaaac1600 in fp_backtrace_many_args at c_call_.c:24
	breakpoint already hit 1 time
....
```
There's probably something slightly wrong with the CFI information being emitted.

Reply:
Thanks for spotting this!

That `CFI_ADJUST` is wrong. I removed it in amd64.S and forgot to when mirroring the changes in arm64.S (I don't know how lldb managed to successfully get a backtrace with the broken CFI entry, though - I tested the qsort.ml program, stopping in `check_frames`).

Reply:
> Not labelling this with `CFI_SIGNAL_FRAME` will mean gdb shows this as a stack frame rather than `<signal handler called>`.

This is true but I'm happy with that. The `CFI_SIGNAL_FRAME` thing is a trick to convince gdb to continue unwinding past a break in stacks, which isn't needed for `caml_c_call_copy_stack_args` since it's called directly on the C stack. I'm happy for it to appear in backtraces, because that means in particular that it shows up in stack-based profilers: copying stack args is rare and somewhat expensive, and so I think we shouldn't hide that it's happening.

Reply:
I agree it's useful to see properly labelled frames.